core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -515,6 +515,10 @@ public CompletableFuture<Optional<Throwable>> acknowledge(String memberId, List<
                             inFlightBatch.maybeInitializeOffsetStateUpdate();
                         }
                         for (Map.Entry<Long, InFlightState> offsetState : inFlightBatch.offsetState.entrySet()) {
+
+                            // updateNextFetchOffset needs to be updated for every record iteration,
+                            // as any record in between could have max delivery count limit exceeded
+                            updateNextFetchOffset = batch.acknowledgeType == AcknowledgeType.RELEASE;
```

Comment:
I think we don't require to check again rather skip the update when offset is marked archived.